contracts/SpokePool.sol

```diff
@@ -146,20 +174,66 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
         numberOfDeposits += 1;
     }
 
-    function initiateRelay(
-        uint256 originChain,
+    function fillRelay(
         address sender,
-        uint256 amount,
         address recipient,
-        uint256 relayerFee,
-        uint256 realizedLpFee
-    ) public {}
-
-    function fillRelay(
-        uint256 relayId,
-        uint256 fillAmount,
+        address destinationToken,
+        uint64 realizedLpFeePct,
+        uint64 relayerFeePct,
+        uint64 depositId,
+        uint256 originChainId,
+        uint256 totalRelayAmount,
+        uint256 maxTokensToSend,
         uint256 repaymentChain
-    ) public {}
+    ) public {
+        // We limit the relay fees to prevent the user spending all their funds on fees.
+        require(
+            relayerFeePct <= 0.5e18 && realizedLpFeePct <= 0.5e18 && (relayerFeePct + realizedLpFeePct) < 1e18,
+            "invalid fees"
+        );
+
+        // Each relay attempt is mapped to the hash of data uniquely identifying it, which includes the deposit data
+        // such as the origin chain ID and the deposit ID, and the data in a relay attempt such as who the recipient
+        // is, which chain and currency the recipient wants to receive funds on, and the relay fees.
+        RelayData memory relayData = RelayData({
+            sender: sender,
+            recipient: recipient,
+            destinationToken: destinationToken,
+            realizedLpFeePct: realizedLpFeePct,
+            relayerFeePct: relayerFeePct,
+            depositId: depositId,
+            originChainId: originChainId,
+            relayAmount: totalRelayAmount
+        });
+        bytes32 relayHash = _getRelayHash(relayData);
+
+        // Compute the equivalent amount to be sent by the relayer before fees have been taken out. This is the amount
+        // that we'll add to the `relayFills` counter, and we do this math here in the contract for the user's
+        // convenience so that they don't have to do this math before calling this function. The user can simply
+        // pass in `maxTokensToSend` and assume that the contract will pull exactly that amount of tokens (or revert).
+        uint256 fillAmountPreFees = (maxTokensToSend / (1e18 - (realizedLpFeePct + relayerFeePct))) * 1e18;
+
+        // Check that the caller is filling a non zero amount of the relay and that the filled amount will not
+        // send to the recipient more than they are expected to receive. Note that the `relays` mapping will point
+        // to the amount filled so far for a particular `relayHash`, so this will start at 0 and increment with each
+        // fill.
+        require(
+            maxTokensToSend > 0 && totalRelayAmount >= relayFills[relayHash] + fillAmountPreFees,
```

Comment:
I think this would fail if the user passes in a number _larger_ than the amount needed to fill the remainder of the relay. Should we do some something like `min(totalRelayAmount - relayFills[relayHash], fillAmountPreFees)` to determine how much they will actually fill?

Reply:
Ah yes i forgot to add this logic, will do

Reply:
Do you think this logic is too convoluted and are there gas savings? https://github.com/across-protocol/across-smart-contracts-v2/pull/9/commits/c6970ae1bd291221333415d20d570cee21dda1c6